torch/onnx/_internal/fx/registration.py

```diff
@@ -0,0 +1,274 @@
+
+    This class stores a collection of overloaded functions that share the same name.
+    Each function is associated with a specific opset version, and multiple versions
+    of the same function can be registered to support different opset versions.
+
+    Attributes:
+        _name: The name of the function group.
+        _functions: A dictionary of functions, keyed by the opset version.
+
+    """
+
+    def __init__(self, name: str) -> None:
+        self._name = name
+        # A dictionary of functions, keyed by the opset version.
+        self._functions: MergeDict[OpsetVersion, Callable] = MergeDict()
+
+    def __repr__(self) -> str:
+        return f"_SymbolicFunctionGroup({self._name}, registered={self._functions})"
+
+    def __getitem__(self, key: OpsetVersion) -> Set[Callable]:
+        result = self.get(key)
+        if result is None:
+            raise KeyError(key)
+        return result
+
+    def get(self, opset: OpsetVersion) -> Optional[Set[Callable]]:
+        """Find the most recent version of the overloaded functions."""
+        return self._functions.get(opset)
+
+    def add(self, func: Callable, opset: OpsetVersion) -> None:
+        """Adds a symbolic function.
+
+        Args:
+            func: The function to add.
+            opset: The opset version of the function to add.
+        """
+        # FIXME(titaiwang): Check if the "function" is ducplicated.
+        self._functions.set_base(opset, func)
+
+    def add_custom(self, func: Callable, opset: OpsetVersion) -> None:
+        """Adds a custom symbolic function.
+
+        Args:
+            func: The symbolic function to register.
+            opset: The corresponding opset version.
+        """
+        self._functions.add_custom(opset, func)
+
+    def remove_custom(self, opset: OpsetVersion) -> None:
+        """Removes a custom symbolic function.
+
+        Args:
+            opset: The opset version of the custom function to remove.
+        """
+        if not self._functions.custom_added(opset):
+            warnings.warn(
+                f"No custom function registered for '{self._name}' opset {opset}"
+            )
+            return
+        self._functions.remove_custom(opset)
+
+    def support_opset(self) -> Collection[OpsetVersion]:
+        """Returns a list of supported opset versions."""
+        return list(self._functions)
+
+
+class OnnxRegistry:
+    """Registry for ONNX functions.
+
+    The registry maintains a mapping from qualified names to symbolic functions,
+    which can be overloaded to support multiple opset versions. It is used to
+    register new symbolic functions and to dispatch calls to the appropriate function.
+
+    Attributes:
+        _registry: A dictionary mapping qualified names to _SymbolicFunctionGroup objects.
+    """
+
+    def __init__(self, opset_version: int = 18) -> None:
+        self._registry: Dict[str, _SymbolicFunctionGroup] = {}
```

Comment:
This dictionary might be too simple to efficiently support multiple opset versions, multiple op overloads in aten and in onnx, custom ops and maybe differentiating prims from aten in the future (not sure whether we will ever support prims, but if there is a chance, maybe we should cover this now)

At https://github.com/pytorch/pytorch/pull/100660/files#diff-ccff3d48c69a764390d32023acc4ba29de4b9cb8b596bd926cb0c6d481990d0cR65 there is a todo saying "# FIXME(titaiwang): How to remove a specific function instead of whole overloads?" which feels is directly associated on how we store the symbolic functions

Reply:
TBH, I don't have much thoughts on registry design, because I think we will definitely have a huge refactor on registry when we decide how to proceed with multiple opset version. On the other hand, to complete dispatcher, I need a draft of registry which leads us to current situation. 

I am not against refactor the registry in this PR, as the dispatcher does engage a lot with it. It's just with people gone for vacation, I don't see more plans on this, and where should we head to with "the design that can make future work easier". We can discuss in the team meeting about this. I am fine even if we need to hold the dispatcher until a great plan on registry is made.